client/src/entrypoints/admin/modal-workflow.test.js

```diff
@@ -0,0 +1,133 @@
+import $ from 'jquery';
+window.$ = $;
+window.jQuery = $;
+
+import '../../../../wagtail/admin/static_src/wagtailadmin/js/vendor/bootstrap-modal';
+import './modal-workflow';
+
+$.get = jest.fn().mockImplementation((url, data, cb) => {
+  cb(JSON.stringify({ html: `<div id="url">${url}</div>`, data, step: 'start' }));
+  return { fail: jest.fn() };
+});
+
+describe('modal-workflow', () => {
+  beforeEach(() => {
+    document.body.innerHTML =
+      '<div id="content"><button class="button action-choose" id="trigger">Open</button></div>';
+  });
+
+  it('exposes module as global', () => {
+    expect(window.ModalWorkflow).toBeDefined();
+  });
+
+  it('should generate a modal when called that is removed when closed', () => {
+    let modalWorkflow;
+
+    const openModal = () => {
+      // eslint-disable-next-line new-cap
+      modalWorkflow = window.ModalWorkflow({ url: 'path/to/endpoint' });
+    };
+
+    expect(document.getElementsByClassName('modal')).toHaveLength(0);
+
+    const triggerButton = document.getElementById('trigger');
+
+    triggerButton.addEventListener('click', openModal);
+    triggerButton.dispatchEvent(new MouseEvent('click'));
+
+    // check that modal has been created in the DOM
+    const modal = document.getElementsByClassName('modal');
+    expect(modal).toHaveLength(1);
+    expect(modal).toMatchSnapshot();
+
+    // check that modalWorkflow returns self
+    expect(modalWorkflow).toBeInstanceOf(Object);
+    expect(modalWorkflow).toHaveProperty('close', expect.any(Function));
```

Comment:
I would say this is superfluous since we’re calling it right after, but at least the error message will be easier to understand.

Reply:
Yeah. This is kind of a habit of mine, first check for presence of the thing and then check the thing has a specific part we want. Makes it easier to debug and also when writing tests. Often I'll remove the first check before committing but forgot here.